Feat/distribution metrics

smglsn12 opened the pull request:
# Add Distribution Level Metrics For Quantitatively Comparing Samples From Two Generative Models

### Add metrics for computing Jensen Shannon distance (for categorical variables), Maximum Mean Discrepancy (for continuous variables) and Fréchet Distance (for MLIP embeddings). 
This PR include an entire restructuring of the processing and base class structure to allow the comparison of two datasets, rather than calculating properties of a specific structure, as the framework has been rigidly built for up until this point.  
Functions are also included for the Shannon entropy of the quantities listed below, but this is not currently integrated into the benchmarks. 

The current functionality can calculate distribution comparisons for:
1. Space group 
2. Crystal System
3. Density (atoms/A^3)
4. Density (g/cm^3)
5. Volume
6. Composition (how many structures contain an element - i.e. CH4 counts as a structure containing carbon and hydrogen)
7. Composition counts (how many times an element appears in the dataset in total - i.e. CH4 counts as 1 carbon and 4 hydrogens)
8. MLIP embeddings for mace, uma, orb, and equiformer - although due to dependency incompatibilities all 4 can't be installed at once. 

### Key Components
Distribution Metrics:

JSDistance: Calculates the Jensen-Shannon distance between two distributions.
MMD: Calculates the Maximum Mean Discrepancy between two distributions. Due to the computation time and memory required, the distributions are random sampled for 1000 samples if they are larger than this.  
FrechetDistance: Calculates the Fréchet Distance between two distributions. 

Implementation Details
JSDistance: 
Space Group, Crystal System - creates a dictionary of space group/crystal system numbers, generates a probability value associated with each quantity, and compares with values computed from a reference dataset (LeMatBulk in this case). 
Composition, Composition Counts - one hot encodes each formula into a set of composite elements/composite element counts. Compares the resulting distribution database wide to that extracted from LeMatBulk. 

MMD:
Density, Volume - computes pairwise distances between two distributions and uses these to determine the Maximum Mean Discrepancy between the two distributions. 

FrechetDistance
MLIP embeddings - computes these from a pymatgen structure object with a specified MLIP (orb, mace, uma, or equiformer) and computes vector means and covariances for the two distributions based on their model embeddings. The distance between these values comprises the FrechetDistance metric. 

Benchmarking: 
This PR also includes a benchmark wrapper class that computes the three metrics for each of the 8 quantities specified above, distilling the output into an individual float for each quantity. See this notebook for an example: (https://github.com/LeMaterial/lematerial-forgebench/blob/feat/distribution-metrics/notebooks/example_distribution_metrics.ipynb)

LeMatBulk Scraping: 
This PR also introduces the significant expansion of the lematbulk scraping directory found here: (https://github.com/LeMaterial/lematerial-forgebench/tree/feat/distribution-metrics/src/lematerial_forgebench/letmat_scraping) which allows the calculation and organization of crystal structure properties currently not available in lematbulk (in this case, most of the quantities being compared here) and the MLIP embeddings. It also includes a script to combine this all into a unified pandas dataframe, which can be used as a reference dataset for the distribution comparison metrics.  

Preprocessing: 
To integrate the distribution metrics with the overall workflow of lemat-forgebench, a preprocessing workflow was developed to turn the sample of structures into a pandas dataframe rather than calculating properties directly on individual structures. This can be found here: (https://github.com/LeMaterial/lematerial-forgebench/blob/feat/distribution-metrics/src/lematerial_forgebench/preprocess/distribution_preprocess.py) The base classes were updated to have optional input variables for the new parameters, such as reference_df, introduced by this change. Providing a reference df is required to use any of the distribution metrics implemented here.    
An additional "unified preprocessor" class was included to hold preprocessing output from the stability preprocessor and the distribution preprocessor, as the outputs of these two preprocessings steps are required for the JSDistance/MMD and FrechetDistance, respectively. This unified preprocessor output holds structure objects with all the resulting information from the two preprocessors as the "property" attribute within each structure object.  

Testing
Given the changes to several base classes, great care was taken to ensure the previous tests are still passing and none of these updates have disrupted the previous workings of the other metrics. Tests have been implemented for the distribution metrics/benchmarks to ensure they successfully find the helper data files and that the changes made to include dataset comparison have been successfully integrated into the entire workflow.

smglsn12 committed:
Feat: add functions for distribution metrics (WIP)

smglsn12 committed:
Feat: add on distribution metrics class

smglsn12 committed:
Merge branch 'main' of github.com:LeMaterial/lematerial-forgebench into feat/distribution-metrics
add updated main

smglsn12 committed:
feat: add Jensen Shannon and MMD distribution metrics and supporting functions/data

smglsn12 committed:
Feat: add frechet dist metric

smglsn12 committed:
Feat/add distribution metrics and benchmark class

smglsn12 committed:
chore: add ruff checks

smglsn12 committed:
fix: fix minor bugs found in linting checks

smglsn12 committed:
fix: addition of reference_df variable to base classes was causing other tests to fail. Added exceptions for this.

smglsn12 committed:
feat: add tests for distribution metrics and benchmarks

smglsn12 committed:
fix: add small toy dataset to distribution tests pass

smglsn12 committed:
fix: add ruff checks

smglsn12 committed:
fix: add helper data files to src/data folder

smglsn12 committed:
fix: add helper data files to src/data folder

smglsn12 committed:
fix minor pathing errors

smglsn12 committed:
refactor: clean up template class for FrechetDistance

smglsn12 committed:
fix ruff checks

vict0rsch reviewed (changes requested):
I'll go further later but we need to clean out comments or explain (inline) why keep them before we can merge

vict0rsch commented on src/lematerial_forgebench/benchmarks/distribution_benchmark.py:
```diff
@@ -0,0 +1,155 @@
+        import math
+
+        def safe_float(value):
+            """Safely convert value to float, handling None and NaN."""
```
What's the intention here with `safe_float`? What is supposed to happen to `None` and `nan`?

vict0rsch committed:
chore: `ruff format`
'15 files reformatted, 39 files left unchanged'

vict0rsch committed:
fix: prevent unexpected execution with `__main__`

vict0rsch committed:
fix: typos

vict0rsch commented on src/lematerial_forgebench/metrics/base.py:
```diff
@@ -432,7 +451,10 @@ def __call__(
         """
         structures_list = format_structures(structures)
-        return self.compute(structures_list)
+        if reference_df is None:
```
Isn't this redundant with the subsequent checks? `self.compute` can handle a `None` `reference_df`

vict0rsch commented on src/lematerial_forgebench/preprocess/distribution_preprocess.py:
```diff
@@ -0,0 +1,73 @@
+
+
+class DistributionPreprocessor(BasePreprocessor):
+    def __init__(
```
This should rather use the native `dataclass` attribute definition interface as above 

```python
DistributionPreprocessor(BasePreprocessor):
    distribution_metric: str = "all"
    ...
```

and move the rest of the logic to `def __post_init__(self):`. Interesting pointer on init vs post_init -> https://stackoverflow.com/questions/68439866/python-dataclass-post-initialisation-for-attributes

smglsn12 commented on src/lematerial_forgebench/benchmarks/distribution_benchmark.py:
```diff
@@ -0,0 +1,155 @@
+            aggregation_method="weighted_mean",
+        )
+
+        # FrechetDistance_metric = FrechetDistance()
```
unlike most of the others - there's actually a reason for this one! I have to turn off the FrechetDistance_metric because it requires the embeddings calculated with PR #21. Once all of these are merged I'll be able to sync that up (but I think this PR should be merged first to prevent merge conflicts)

vict0rsch replied:
Sounds good, I'm de-commenting because we merged #21 here

smglsn12 commented on src/lematerial_forgebench/benchmarks/distribution_benchmark.py:
```diff
@@ -0,0 +1,155 @@
+            )
+
+        # FrechetDistance_results = evaluator_results.get("FrechetDistance")
+        # if FrechetDistance_results:
```
same as the comment above!

sid-betalol committed:
feature: add support for more MLIPs (#21)

* Initial empty commit

* feat: add support for new MLIPs for energy, embeddings, relaxers

* fix: dependencies between models

* fix: main check to avoid extras

* fix: fix dependency issues and bugs

* wip embeddings and calc

* Refactor/registry (#30)

* refactor: systematic registry discovery and config building

* feat: rich tables for model info

* docs: add docstrings and module description

* chore: UMA name

---------

Co-authored-by: Ali Ramlaoui <ramlaouiali@gmail.com>

* chore: Clean universal stability preprocess and remove base stability preprocessor

* fix: Remove no_grad for energy conserving models

* chore: Remove previous stability preprocessor

* chore: Add extract_embeddings arg

* fix: model_type being used two times as arg of create function

* chore: Unused arg

* feat: Universal stability with config in cli

* test: Pass stability tests

* fix: Use orb and mace together

* chore: Clean unused args and classes

* chore: stress not optional

* docs: help to find available models

* chore: readability of dict gets

* fix: uma transform

---------

Co-authored-by: Ali Ramlaoui <ramlaouiali@gmail.com>
Co-authored-by: Andy Xu <andxu@teapot.cs.hmc.edu>
Co-authored-by: Samuel Gleason <50921084+smglsn12@users.noreply.github.com>
Co-authored-by: Victor Schmidt <9283470+vict0rsch@users.noreply.github.com>

vict0rsch committed:
fix: broken lock -> to check

vict0rsch committed:
refactor: improve docs & readability

vict0rsch committed:
refactor: use `min` instead of control flow
Iterating over the dataset we can use `min(len(dataset), i + 100_000)` instead of +35_299

vict0rsch committed:
fix: remove trailing comments

vict0rsch committed:
chore: format toml

vict0rsch committed:
refactor: centralize `safe_float`
even if it's a no-op

vict0rsch committed:
fix: Don't ignore import sorting

vict0rsch committed:
chore: ignore notebooks in import checks

vict0rsch committed:
chore: remove unused imports

vict0rsch committed:
wip: towards a unified API for metrics

smglsn12 committed:
fix: reference_df inconsistent API calls

smglsn12 committed:
fix: run ruff checks

Ramlaoui committed:
chore: simple docstring to retrigger code action

Ramlaoui committed:
chore: remove breakpoint

Ramlaoui commented on src/lematerial_forgebench/metrics/distribution_metrics.py:
```diff
@@ -0,0 +1,389 @@
+
+    @staticmethod
+    def compute_structure(structure: pd.DataFrame, **compute_args: Any) -> dict:
+        """Compute the similarity of the structure to a target distribution.
```
There is a misunderstanding in how the `BaseMetric` class works. `compute_structure` is usually used only for one structure, and there is a specific method that gets called with all the structures called `compute`. I suggest overriding `compute` itself to be compatible with the API, which would also be 100% compatible with this particular benchmark

Ramlaoui commented on src/lematerial_forgebench/metrics/distribution_metrics.py:
```diff
@@ -0,0 +1,389 @@
+        float
+            MMD
+        """
+        np.random.seed(32)
```
Can we make the seed a class attribute?

Ramlaoui commented on src/lematerial_forgebench/metrics/distribution_metrics.py:
```diff
@@ -0,0 +1,389 @@
+        dist_metrics = {}
+        for quant in quantities:
+            if quant in reference_df.columns:
+                if isinstance(reference_df[quant].iloc[0], np.float64):
```
Worth a comment to say that the JS metric measures divergence between discrete properties and not continuous ones

Ramlaoui commented on src/lematerial_forgebench/utils/distribution_utils.py:
```diff
@@ -0,0 +1,157 @@
+        for row in strut_list:
+            prob_dict[str(int(row[0]))] = row[1]
+
+    if metric_type == np.ndarray:
+        prob_dict = {}
+        for i in range(0, 118):
+            prob_dict[str(i + 1)] = 0
+
+        one_hots = np.zeros(118)
+        for i in range(0, len(df)):
+            one_hots += df.iloc[i][metric]
+
+        one_hots = one_hots / sum(one_hots)
+
+        for i in range(0, 118):
+            prob_dict[str(i + 1)] = one_hots[i]
```
There's probably a needed verification here that the metric passed is atomic composition (which I assume because of the 118, but this is not obvious from the code).

Moreover, this function gets called for the Density metric in my example which I assume is not expected?

Ramlaoui commented on src/lematerial_forgebench/benchmarks/distribution_benchmark.py:
```diff
@@ -0,0 +1,174 @@
+            aggregation_method="weighted_mean",
+        )
+
+        # Initialize the MFrechetDistanceMD metric
+        # FrechetDistance_metric = FrechetDistance(reference_df=reference_df)
+
+        # # add to evaluator config
+        # evaluator_configs["FrechetDistance"] = EvaluatorConfig(
+        #     name="FrechetDistance Analysis",
+        #     description="Calculates the Frechet Distance between two distributions",
+        #     metrics={"FrechetDistance": FrechetDistance_metric},
+        #     weights={"FrechetDistance": 1.0},
+        #     aggregation_method="weighted_mean",
+        # )
```
Is there a reason why this is commented (might have already been adressed in someone else's, feel free to ignore if that's the case)

Ramlaoui commented on src/lematerial_forgebench/benchmarks/base.py:
```diff
@@ -137,6 +140,7 @@ def aggregate_evaluator_results(
     def evaluate(
         self,
         structures: list[Structure],
+        reference_df: pd.DataFrame | None = None,
     ) -> BenchmarkResult:
```
With the `_compute_args` method abstraction, there is no need to hardcode such arguments in this function, we can simply get them from calling the method

Ramlaoui committed:
chore: Fix distribution metrics API compatibility

Ramlaoui committed:
chore: preprocessor returns structures instead of dataframe

Ramlaoui committed:
chore: Revert base evaluator API

Ramlaoui committed:
tests: Adapt tests to new interface

Ramlaoui committed:
tests: Adapt tests to new interface

Ramlaoui commented:
I also have an additional conceptual suggestion:

In the current setting, one metric is defined as a method (for example Jensen-Shannon Divergence is just a tool to measure distances between distributions, but it is used as a metric measurement). This leads to having one metric = multiple measures of different characteristics of the generated structures (Space Group, Composition distributions etc.).

A more unified approach would be to have a base class for Jensen-Shannon Divergence measurements, but which defines several submetrics, each returning one measurement.

smglsn12 committed:
Feat: add example code for distribution metrics individually

smglsn12 committed:
update frechet distance to interface with MLIPs

smglsn12 committed:
add scripts to generate embeddings for lematbulk

smglsn12 committed:
Refactor: rename lematbulk.py to be more descriptive

smglsn12 committed:
Chore: run ruff checks

smglsn12 committed:
Feat: add distribution benchmarks

smglsn12 committed:
Feat: update tests for distribution benchmark to include Frechet Distance

smglsn12 committed:
Chore: update gitignore to include more filetypes

smglsn12 committed:
Fix: update example notebook

smglsn12 committed:
Feat: add function timeout option for stability calculations

smglsn12 committed:
Feat: develop script for bulk calculations of stability preprocessor data for all LeMatBulk

smglsn12 committed:
Fix: import bugs for equiformer

smglsn12 committed:
chore: run ruff checks

smglsn12 committed:
Fix: node embeddings not storing correctly

smglsn12 committed:
Fix: dataset length returned to full

Ramlaoui committed:
chore: Support batched embeddings computation

smglsn12 committed:
Feat: add script to combine sub data frames into a full dataframe

smglsn12 committed:
Chore: remove half finished pca from distribution metrics

Ramlaoui committed:
chore: remove no longer computed keys

Ramlaoui committed:
chore: torch geometric dependency

smglsn12 committed:
Feat: update Frechet distance metric for increased flexibility with different mlips and add example to distribution_metrics.py

smglsn12 committed:
Feat: add tests for Frechet Distance and restructure datascheme so one unified reference df can be passed for each distribution metric when calling distribution benchmark

smglsn12 committed:
chore: run ruff checks

smglsn12 committed:
refactor: update example distribution notebook

smglsn12 committed:
fix: add important data files

smglsn12 commented:
I suggest merging this branch into https://github.com/LeMaterial/lematerial-forgebench/tree/feat/diversity-metric - this branch is pretty behind at this point and will leverage a lot of the new functionality in the preprocessors 
cc @Ramlaoui @vict0rsch @sid-betalol

vict0rsch committed:
Merge branch 'main' into feat/distribution-metrics

vict0rsch commented on .gitignore:
```diff
@@ -172,4 +172,9 @@ src/lematerial_forgebench/.DS_Store
+*.npy
+*.pkl
+*.cif
+*.json
```
This looks weird to me given the amount of `.json` data in this PR

smglsn12 committed:
refactor: Remove .json files from .gitignore

smglsn12 committed:
fix: add important data file

smglsn12 committed:
fix: remove .json from git ignore

smglsn12 committed:
chore: run ruff checks

smglsn12 committed:
fix: update distribution metrics tests

smglsn12 committed:
chore: run ruff checks

vict0rsch reviewed (approved):
Thank you for your hard work @smglsn12 👌 there are things that will keep evolving but let's iterate and not aim for perfect so it's merging time !!
